File: src/main/java/io/femo/http/transport/http2/frames/PushPromiseFrame.java
package io.femo.http.transport.http2.frames;

import io.femo.http.Constants;
import io.femo.http.transport.http2.HttpFrame;
import io.femo.http.transport.http2.HttpFrameException;
import io.femo.http.transport.http2.HttpSettings;
import io.femo.http.transport.http2.HttpUtil;

import java.nio.ByteBuffer;

/**
 * Created by felix on 9/18/16.
 */
public class PushPromiseFrame extends PaddedFrame {

    private int promisedStreamId;
    private byte[] headerBlockFragment;

    /**
     * Constructs a new HttpFrame with no payload and information
     *
     * @param settings the settings for the current HTTP/2,0 connection
     */
    public PushPromiseFrame(HttpSettings settings) {
        super(settings);
        setType(Constants.Http20.FrameType.PUSH_PROMISE);
    }

    public boolean isEndHeaders() {
        return (getFlags() & 0x4) == 0x4;
    }

    public void setEndHeaders(boolean endHeaders) {
        setFlags((short) (getFlags() & ~0x4 | (endHeaders ? 0x4 : 0)));
    }

    @Override
    public void build() {
        int length = headerBlockFragment.length + 4;    //stream identifier length
        if(isPadded()) {
            length++; //padding adds one byte
            length += getPadLength();
        }
        ByteBuffer buffer = ByteBuffer.allocate(length);
        if(isPadded())
            buffer.put(HttpUtil.toByte(getPadLength()), 1, 1);
        buffer.putInt(promisedStreamId);
        buffer.put(headerBlockFragment);
        setPayload(buffer.array());
    }

    public int getPromisedStreamId() {
        return promisedStreamId;
    }

    public void setPromisedStreamId(int promisedStreamId) {
        this.promisedStreamId = promisedStreamId;
    }

    public byte[] getHeaderBlockFragment() {
        return headerBlockFragment;
    }

    public void setHeaderBlockFragment(byte[] headerBlockFragment) {
        this.headerBlockFragment = headerBlockFragment;
    }

    public static PushPromiseFrame from(HttpFrame frame) {
        PushPromiseFrame pushPromiseFrame = new PushPromiseFrame(frame.getSettings());
        pushPromiseFrame.setFlags(frame.getFlags());
        pushPromiseFrame.setStreamIdentifier(frame.getStreamIdentifier());
        ByteBuffer byteBuffer = ByteBuffer.wrap(frame.getPayload());
        if(pushPromiseFrame.isPadded()) {
            pushPromiseFrame.setPadLength((short) (byteBuffer.get() & 0xff));
        }
        pushPromiseFrame.setPromisedStreamId(byteBuffer.getInt());
        byte[] headerBlockFragment = new byte[byteBuffer.remaining() - pushPromiseFrame.getPadLength()];
        byteBuffer.get(headerBlockFragment);
        pushPromiseFrame.setHeaderBlockFragment(headerBlockFragment);
        if(byteBuffer.remaining() != pushPromiseFrame.getPadLength()) {
            throw new HttpFrameException("Invalid frame padding length");
        }
        return pushPromiseFrame;
    }
}

File: src/main/java/org/web25/http/transport/http2/frames/PushPromiseFrame.kt
package org.web25.http.transport.http2.frames

import org.web25.http.Constants
import org.web25.http.transport.http2.HttpFrame
import org.web25.http.transport.http2.HttpFrameException
import org.web25.http.transport.http2.HttpSettings
import org.web25.http.transport.http2.HttpUtil

import java.nio.ByteBuffer

/**
 * Created by felix on 9/18/16.
 */
class PushPromiseFrame
/**
 * Constructs a new HttpFrame with no payload and information

 * @param settings the settings for the current HTTP/2,0 connection
 */
(settings: HttpSettings) : PaddedFrame(settings) {

    var promisedStreamId: Int = 0
    var headerBlockFragment: ByteArray? = null

    init {
        type = Constants.Http20.FrameType.PUSH_PROMISE
    }

    var isEndHeaders: Boolean
        get() = flags.toInt() and 0x4 == 0x4
        set(endHeaders) {
            flags = (flags.toInt() and 0x4.inv() or if (endHeaders) 0x4 else 0).toShort()
        }

    override fun build() {
        var length = headerBlockFragment!!.size + 4    //stream identifier length
        if (isPadded) {
            length++ //padding adds one byte
            length += padLength.toInt()
        }
        val buffer = ByteBuffer.allocate(length)
        if (isPadded)
            buffer.put(HttpUtil.toByte(padLength), 1, 1)
        buffer.putInt(promisedStreamId)
        buffer.put(headerBlockFragment!!)
        payload = buffer.array()
    }

    companion object {

        fun from(frame: HttpFrame): PushPromiseFrame {
            val pushPromiseFrame = PushPromiseFrame(frame.settings)
            pushPromiseFrame.flags = frame.flags
            pushPromiseFrame.streamIdentifier = frame.streamIdentifier
            val byteBuffer = ByteBuffer.wrap(frame.payload)
            if (pushPromiseFrame.isPadded) {
                pushPromiseFrame.padLength = (byteBuffer.get().toInt() and 0xff).toShort()
            }
            pushPromiseFrame.promisedStreamId = byteBuffer.int
            val headerBlockFragment = ByteArray(byteBuffer.remaining() - pushPromiseFrame.padLength)
            byteBuffer.get(headerBlockFragment)
            pushPromiseFrame.headerBlockFragment = headerBlockFragment
            if (byteBuffer.remaining() != pushPromiseFrame.padLength.toInt()) {
                throw HttpFrameException("Invalid frame padding length")
            }
            return pushPromiseFrame
        }
    }
}
